File: sdk/node/src/signing.ts
import { createHmac } from 'crypto';

/**
 * Computes the `X-Signature`/`X-Timestamp` pair `HmacSignatureGuard`
 * verifies — `${timestamp}.${method}.${path}.${rawBody}`, HMAC-SHA256,
 * hex digest (see docs/guide/api/README.md#hmac-request-signing).
 * `path` must be the exact request path the server sees, including the
 * `/api/v1` prefix and query string if any — the guard signs
 * `request.originalUrl`, not a normalized or query-stripped version.
 * `body` must be the *exact* bytes sent on the wire (this SDK always
 * signs the same JSON string it then sends, never a value re-serialized
 * afterward — see `HmacSignatureGuard`'s own docblock for why
 * re-serialization would silently break signatures for some payloads).
 */
export function signRequest(
  secret: string,
  method: string,
  path: string,
  body: string,
): { signature: string; timestamp: string } {
  const timestamp = String(Math.floor(Date.now() / 1000));
  const signedPayload = `${timestamp}.${method.toUpperCase()}.${path}.${body}`;
  const signature = createHmac('sha256', secret).update(signedPayload).digest('hex');
  return { signature, timestamp };
}

File: sdk/dotnet/src/Signing.cs
using System.Security.Cryptography;
using System.Text;

namespace OmniSwitch.Sdk;

/// <summary>
/// Computes the <c>X-Signature</c>/<c>X-Timestamp</c> pair
/// <c>HmacSignatureGuard</c> verifies server-side:
/// <c>HMAC-SHA256(secret, "${timestamp}.${method}.${path}.${rawBody}")</c>,
/// hex digest. <paramref name="path"/> must be the exact request path the
/// server sees, including the <c>/api/v1</c> prefix and query string if
/// any — the guard signs <c>request.originalUrl</c>, not a normalized or
/// query-stripped version. <paramref name="body"/> must be the exact
/// bytes sent on the wire — <see cref="OmniSwitchClient"/> always signs
/// the same JSON string it then sends, never a value re-serialized
/// afterward.
/// </summary>
public static class Signing
{
    public readonly record struct SignResult(string Signature, string Timestamp);

    public static SignResult SignRequest(string secret, string method, string path, string body)
    {
        var timestamp = DateTimeOffset.UtcNow.ToUnixTimeSeconds().ToString();
        var signedPayload = $"{timestamp}.{method.ToUpperInvariant()}.{path}.{body}";
        var signature = HmacSha256Hex(secret, signedPayload);
        return new SignResult(signature, timestamp);
    }

    internal static string HmacSha256Hex(string secret, string payload)
    {
        using var hmac = new HMACSHA256(Encoding.UTF8.GetBytes(secret));
        var hash = hmac.ComputeHash(Encoding.UTF8.GetBytes(payload));
        return Convert.ToHexString(hash).ToLowerInvariant();
    }
}
